### PR TITLE
Driver: Include non-surplus-capturing-jit-orders in the driver /solve

### LINKED ISSUE #3004: feat: Report and store all executed orders during bidding
### Problem
At the moment, a solver can include orders in a settlement at will if the order is settled at limit price. Those orders need not be reported during bidding.

This is generally a problem for protecting users but will also be problematic with most forms of combinatorial auction or payments with multiple winners. For example, a solver might execute an additional order on a token pair having no tokens in common with the reported trades. A second winner might now execute on that same token pair.

### Suggested solution
A first step towards creating more rigorous rules around settling orders out of competition would be to force solvers to report all orders which are executed to the autopilot.

### Additional context
There is a general problem with executing orders which can capture surplus at limit price.

The feature of executing jit orders at limit price without reporting the order is only used in obscure cases. For example for CoW AMM JIT orders with missing native prices. The impact of removing the feature should be small.

### PR DESCRIPTION
**Pull request by mateo-mro**
# Description
Include the non-surplus-capturing JIT orders in the driver `/solve` response. For more context, please see https://github.com/cowprotocol/services/issues/3004

# Changes
- Include non-surplus-capturing JIT orders in the driver `/solve` response
- Do not use the clearing prices for such orders, but the  JIT order executed priced (since it is also used in the encoding)

## How to test
1. Driver tests

## Related Issues

Fixes #3004

### CODE REVIEW

**Comment by sunce86** [comment]
Added time-to-happy-moo label as this fix is important for proper selection of winners in autopilot.

**Review by MartinquaXD — COMMENTED** [review]
AFAICS the issue did not mention to alter the meaning of what it means for a solution to be non-empty which seems like the source of at least a portion of the diff.
My understanding is that the only goal is to make solvers report JIT orders even if they don't increase the score and not propagate solutions with potentially 0 score.

File: crates/driver/src/domain/competition/solution/mod.rs [hunk 1]
```diff
@@ -267,11 +267,14 @@ impl Solution {
 
     /// An empty solution has no trades which is allowed to capture surplus and
     /// a score of 0.
-    pub fn is_empty(&self, surplus_capturing_jit_order_owners: &HashSet<eth::Address>) -> bool {
-        !self
-            .trades
-            .iter()
-            .any(|trade| self.trade_count_for_scorable(trade, surplus_capturing_jit_order_owners))
+    pub fn is_empty(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/mod.rs:270]: Score for pure JIT orders is 0, and solutions with zero score are rejected by autopilot automatically (this is a requirement from one of the CIPs). I'd agree that the definition of empty solutions shouldn't change.

File: crates/driver/src/domain/competition/solution/settlement.rs [hunk 2]
```diff
@@ -255,29 +260,49 @@ impl Settlement {
...
+                        sell: trade.sell(),
+                        buy: trade.buy(),
+                        executed_sell: match trade.side() {
+                            Side::Buy => trade.buy().amount,
+                            Side::Sell => trade.executed().into(),
+                        },
+                        executed_buy: match trade.side() {
+                            Side::Buy => trade.executed().into(),
+                            Side::Sell => trade.sell().amount,
+                        },
```
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/settlement.rs:301]: Reference: see how the sell and buy amounts are calculated based on custom prices: https://github.com/cowprotocol/services/blob/main/crates/driver/src/domain/competition/solution/scoring.rs#L197-L209

And for JIT orders, custom prices are:
```
sell = buy_amount
buy = sell_amount
```

**Comment by squadgazzz** [comment]
Would it be possible to specify the owners list while building a fake auction for the quotes?
https://github.com/cowprotocol/services/blob/8cbe5ecc239d542cf295fa0421fccca46d687ccc/crates/driver/src/domain/quote.rs#L186
This is required for another PR(https://github.com/cowprotocol/services/pull/3038) to avoid some ugly workaround.

File: crates/driver/src/domain/competition/solution/settlement.rs [hunk 3]
```diff
@@ -261,48 +261,34 @@ impl Settlement {
...
-            }
+            let prices = match trade {
+                Trade::Fulfillment(_) => ClearingPrices {
+                    sell: self.solution.prices[&trade.sell().token.wrap(self.solution.weth)],
+                    buy: self.solution.prices[&trade.buy().token.wrap(self.solution.weth)],
+                },
+                Trade::Jit(_) => ClearingPrices {
+                    sell: trade.buy().amount.into(),
+                    buy: trade.sell().amount.into(),
+                },
```
Comment (sunce86) [inline crates/driver/src/domain/competition/solution/settlement.rs:272]: I don't think this is equivalent to what I suggested.

`trade.sell_amount()` and `trade.buy_amount()` functions are supposed to receive `uniform` prices (prices reported by solvers API), while here you are building `custom` prices (that include the `fee` from JIT order). 

Can you double check this with some unit/driver tests?

**Comment by squadgazzz** [comment]
> Yeah, definitely, but it is totally unrelated to this PR. I can address it in another PR if you want 👌

Ok, NVM, I'll rebase my PR once this one is merged.

File: crates/driver/src/tests/setup/mod.rs [hunk 4]
```diff
@@ -1224,6 +1224,42 @@ impl<'a> SolveOk<'a> {
...
+        .unwrap();
+
+        // Since JIT orders don't have UID at creation time, we need to search for
+        // matching token pair
+        for expected in jit_orders.iter() {
+            for trade in trades.values() {
+                let u256 = |value: &serde_json::Value| {
+                    eth::U256::from_dec_str(value.as_str().unwrap()).unwrap()
+                };
+                let token = |value: &serde_json::Value| value.as_str().unwrap().to_string();
```
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/driver/src/tests/setup/mod.rs:1243]: Why do we need this closure here? Looks like just a redundant line of code.

File: crates/driver/src/domain/competition/solution/settlement.rs [hunk 5]
```diff
@@ -254,34 +259,86 @@ impl Settlement {
...
+                            .unwrap_or_else(|err| log_err(trade, err, "buy_amount")),
+                    };
+                    acc.insert(trade.uid(), order);
+                }
+            }
         }
         acc
     }
 
+    fn jit_order_executed_buy(trade: &Trade) -> Result<eth::TokenAmount, error::Math> {
```
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/driver/src/domain/competition/solution/settlement.rs:307]: Why is it called `jit_order_executed_buy,` while internally, it also works with the regular Fulfillment?

Doesn't existing function produce the same result?
https://github.com/cowprotocol/services/blob/8cbe5ecc239d542cf295fa0421fccca46d687ccc/crates/driver/src/domain/competition/solution/trade.rs#L71-L109

File: crates/driver/src/domain/competition/solution/settlement.rs [hunk 6]
```diff
@@ -254,30 +259,49 @@ impl Settlement {
...
-                }),
-                executed_buy: trade.buy_amount(&prices).unwrap_or_else(|err| {
-                    // This should never happen, returning 0 is better than panicking, but we
-                    // should still alert.
-                    tracing::error!(?trade, prices=?self.solution.prices, ?err, "could not compute buy_amount");
-                    0.into()
-                }),
-            };
-            acc.insert(trade.order().uid, order);
+        for trade in self.solution.market_trades() {
```
Comment (sunce86) [inline crates/driver/src/domain/competition/solution/settlement.rs:270]: Hm, TBH I would not filter anything here. I think the point it so just report all types of orders. So, just iterate over all trades from solution 

@MartinquaXD WDYT

**Review by sunce86 — COMMENTED** [review]
Almost there

File: crates/driver/openapi.yml [hunk 7]
```diff
@@ -237,7 +237,7 @@ components:
           enum: ["erc20", "internal"]
         class:
           type: string
-          enum: ["market", "limit", "liquidity"]
+          enum: ["market", "limit"]
```
Comment (sunce86) [inline crates/driver/openapi.yml:240]: 🙏

File: crates/driver/src/domain/competition/solution/settlement.rs [hunk 8]
```diff
@@ -254,30 +259,49 @@ impl Settlement {
...
+                        sell: trade.sell(),
+                        buy: trade.buy(),
+                        executed_sell: trade
+                            .sell_amount(&prices)
+                            .unwrap_or_else(|err| log_err(trade, err, "sell_amount")),
+                        executed_buy: trade
+                            .buy_amount(&prices)
+                            .unwrap_or_else(|err| log_err(trade, err, "buy_amount")),
+                    };
+                    acc.insert(trade.uid(), order);
```
Comment (sunce86) [inline crates/driver/src/domain/competition/solution/settlement.rs:288]: nit: `acc.insert(trade.uid(), order);` can be extracted for both cases

**Review by sunce86 — APPROVED** [review]
LGTM. Nice

File: crates/driver/src/tests/setup/mod.rs [hunk 9]
```diff
@@ -1224,6 +1224,41 @@ impl<'a> SolveOk<'a> {
...
+            for trade in trades.values() {
+                let u256 = |value: &serde_json::Value| {
+                    eth::U256::from_dec_str(value.as_str().unwrap()).unwrap()
+                };
+                let sell_token = trade.get("sellToken").unwrap().to_string();
+                let buy_token = trade.get("buyToken").unwrap().to_string();
+
+                if sell_token == expected.order.sell_token && buy_token == expected.order.buy_token
+                {
+                    assert!(is_approximately_equal(
```
Comment (MartinquaXD) [inline crates/driver/src/tests/setup/mod.rs:1248]: Is `is_approximately_equal` actually necessary here?
Don't we know for sure which values to expect exactly?

File: crates/driver/src/tests/setup/mod.rs [hunk 10]
```diff
@@ -1224,6 +1224,41 @@ impl<'a> SolveOk<'a> {
...
+                        expected.order.sell_amount,
+                        u256(trade.get("executedSell").unwrap())
+                    ));
+                    assert!(is_approximately_equal(
```
Comment (MartinquaXD) [inline crates/driver/src/tests/setup/mod.rs:1252]: Using asserts here means we could not have multiple JIT orders trading the same tokens with different amounts.
I would rather expect some code that calls `trades.values().find()` where `find()` returns true if all relevant values match and afterwards we assert that for all expected trades some jit order could be found.

File: crates/driver/src/tests/setup/solver.rs [hunk 11]
```diff
@@ -270,18 +270,25 @@ impl Solver {
...
-                                .blockchain
-                                .get_token_wrapped(jit.quoted_order.order.buy_token),
-                            (jit.execution.sell - jit.quoted_order.order.surplus_fee()).to_string(),
-                        );
+                        // Skipping the prices for JIT orders (non-surplus-capturing)
+                        if config
+                            .expected_surplus_capturing_jit_order_owners
+                            .contains(&jit.quoted_order.order.owner)
+                        {
+                            prices_json.insert(
```
Comment (MartinquaXD) [inline crates/driver/src/tests/setup/solver.rs:278]: Probably makes sense to assert that `.insert()` returned `None` to ensure that we don't overwrite prices for existing tokens by accident. (also applies to the code handling regular orders).

File: crates/driver/src/domain/competition/solution/settlement.rs [hunk 12]
```diff
@@ -254,30 +259,46 @@ impl Settlement {
...
+                        sell: self.solution.prices[&trade.sell().token.wrap(self.solution.weth)],
+                        buy: self.solution.prices[&trade.buy().token.wrap(self.solution.weth)],
+                    };
+                    competition::Amounts {
+                        side: trade.side(),
+                        sell: trade.sell(),
+                        buy: trade.buy(),
+                        executed_sell: trade
+                            .sell_amount(&prices)
+                            .unwrap_or_else(|err| log_err(trade, err, "sell_amount")),
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/settlement.rs:283]: Should these err for `executed_sell` and `executed_buy`?

**Review by MartinquaXD — APPROVED** [review]

File: crates/driver/src/tests/setup/mod.rs [hunk 13]
```diff
@@ -1224,6 +1225,45 @@ impl<'a> SolveOk<'a> {
...
+        .unwrap();
+
+        // Since JIT orders don't have UID at creation time, we need to search for
+        // matching token pair
+        for expected in jit_orders.iter() {
+            let mut exist = false;
+            for trade in trades.values() {
+                exist |= self.exist_jit_order(trade, expected)
+            }
+            assert!(exist, "JIT order {expected:?} not found");
```
Comment (sunce86) [inline crates/driver/src/tests/setup/mod.rs:1244]: nit: the construction seems too complex. Why don't you just:

```
for expected in jit_orders.iter() {
   for trade in trades.values() {
      let exist = self.exist_jit_order(trade, expected);
      assert!(exist, "JIT order {expected:?} not found");
   }
}
```

**Review by sunce86 — APPROVED** [review]

File: crates/driver/src/tests/setup/mod.rs (hunk 13 above)
Comment (MartinquaXD) [reply crates/driver/src/tests/setup/mod.rs:1244]: Why not:
```rust
        for expected in jit_orders.iter() {
            let exists = trades.values().any(|trade| self.trade_matches(trade, expected));
            assert!(exists, "JIT order {expected:?} not found");
        }
```

File: crates/driver/src/tests/setup/solver.rs [hunk 14]
```diff
@@ -197,20 +197,24 @@ impl Solver {
...
-                        );
-                        prices_json.insert(
-                            config
-                                .blockchain
-                                .get_token_wrapped(fulfillment.quoted_order.order.buy_token),
-                            (fulfillment.execution.sell
-                                - fulfillment.quoted_order.order.surplus_fee())
-                            .to_string(),
-                        );
+                        assert!(prices_json
```
Comment (MartinquaXD) [inline crates/driver/src/tests/setup/solver.rs:200]: nit: Having such big expressions is hard to read and less expressive than properly named local variables.
```rust
                        let previous_value = prices_json
                            .insert(
                                config
                                    .blockchain
                                    .get_token_wrapped(fulfillment.quoted_order.order.sell_token),
                                fulfillment.execution.buy.to_string(),
                            );
                        assert_eq!(previous_value, None, "existing price overwritten");
```

**Pull request merged**